core/src/layers/foyer.rs

```diff
@@ -0,0 +1,420 @@
+        async move {
+            let (rp, w) = self.inner.accessor.write(path, args).await?;
+            Ok((
+                rp,
+                Writer {
+                    w,
+                    q: Some(QueueBuf::new()),
+                    path: path.to_string(),
+                    inner,
+                },
+            ))
+        }
+    }
+
+    fn delete(&self) -> impl Future<Output = Result<(RpDelete, Self::Deleter)>> + MaybeSend {
+        let inner = self.inner.clone();
+        async move {
+            let (rp, d) = inner.accessor.delete().await?;
+            Ok((
+                rp,
+                Deleter {
+                    d,
+                    keys: vec![],
+                    inner,
+                },
+            ))
+        }
+    }
+
+    fn list(
+        &self,
+        path: &str,
+        args: OpList,
+    ) -> impl Future<Output = Result<(RpList, Self::Lister)>> + MaybeSend {
+        self.inner.accessor.list(path, args)
+    }
+
+    fn info(&self) -> std::sync::Arc<AccessorInfo> {
+        self.inner.accessor.info()
+    }
+
+    fn create_dir(
+        &self,
+        path: &str,
+        args: OpCreateDir,
+    ) -> impl Future<Output = Result<RpCreateDir>> + MaybeSend {
+        self.inner.accessor.create_dir(path, args)
+    }
+
+    fn copy(
+        &self,
+        from: &str,
+        to: &str,
+        args: OpCopy,
+    ) -> impl Future<Output = Result<RpCopy>> + MaybeSend {
+        // TODO(MrCroxx): Implement copy with foyer cache.
+        self.inner.accessor.copy(from, to, args)
+    }
+
+    fn rename(
+        &self,
+        from: &str,
+        to: &str,
+        args: OpRename,
+    ) -> impl Future<Output = Result<RpRename>> + MaybeSend {
+        // TODO(MrCroxx): Implement copy with foyer cache.
+        self.inner.accessor.rename(from, to, args)
+    }
+
+    fn stat(&self, path: &str, args: OpStat) -> impl Future<Output = Result<RpStat>> + MaybeSend {
+        // TODO(MrCroxx): Implement copy with foyer cache.
+        self.inner.accessor.stat(path, args)
+    }
+
+    fn presign(
+        &self,
+        path: &str,
+        args: OpPresign,
+    ) -> impl Future<Output = Result<RpPresign>> + MaybeSend {
+        self.inner.accessor.presign(path, args)
+    }
+}
+
+pub struct Writer<A: Access> {
+    w: A::Writer,
+    q: Option<QueueBuf>,
+    path: String,
+    inner: Arc<Inner<A>>,
+}
+
+impl<A: Access> oio::Write for Writer<A> {
+    async fn write(&mut self, bs: Buffer) -> Result<()> {
+        self.q.as_mut().unwrap().push(bs.clone());
+        self.w.write(bs).await
+    }
+
+    async fn close(&mut self) -> Result<Metadata> {
+        let buffer = self.q.take().unwrap().collect();
+        let res = self.w.close().await;
+        self.inner.cache.insert(self.path.clone(), buffer.into());
+        res
+    }
+
+    async fn abort(&mut self) -> Result<()> {
+        self.w.abort().await
+    }
+}
+
+pub struct Deleter<A: Access> {
+    d: A::Deleter,
```

Comment:
I suggest we avoid saving word here. Writing `deleter` in the IDE isn't much trouble, but it can make the code harder to read.